controllers/medusa/medusarestorejob_controller.go

```diff
@@ -0,0 +1,294 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package medusa
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/google/uuid"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	medusav1alpha1 "github.com/k8ssandra/k8ssandra-operator/apis/medusa/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/config"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/medusa"
+)
+
+// MedusaRestoreJobReconciler reconciles a MedusaRestoreJob object
+type MedusaRestoreJobReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme *runtime.Scheme
+	medusa.ClientFactory
+}
+
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusarestorejobs,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusarestorejobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusabackups,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cassandra.datastax.com,namespace="k8ssandra",resources=cassandradatacenters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=apps,namespace="k8ssandra",resources=statefulsets,verbs=list;watch
+
+func (r *MedusaRestoreJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("medusarestorejob", req.NamespacedName)
+	factory := medusa.NewFactory(r.Client, logger)
+	logger.Info("Starting reconcile", "MedusaRestoreJob", req.NamespacedName.Name)
+	request, result, err := factory.NewMedusaRestoreRequest(ctx, req.NamespacedName)
+
+	if result != nil {
+		return *result, err
+	}
+
+	if request.RestoreJob.Status.StartTime.IsZero() {
+		request.SetMedusaRestoreStartTime(metav1.Now())
+		request.SetMedusaRestoreKey(uuid.New().String())
+		request.SetMedusaRestorePrepared(false)
+	}
+
+	cassdcKey := types.NamespacedName{Namespace: request.RestoreJob.Namespace, Name: request.Datacenter.Name}
+	cassdc := &cassdcapi.CassandraDatacenter{}
+	err = r.Get(ctx, cassdcKey, cassdc)
+	if err != nil {
+		logger.Error(err, "failed to get cassandradatacenter", "CassandraDatacenter", cassdcKey)
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	// Prepare the restore by placing a mapping file in the Cassandra data volume.
+	if !request.RestoreJob.Status.RestorePrepared {
```

Comment:
We don't need this for an in-place restore do we? I haven't tested but if I recall I want to say it is fine to do; however, that is a good bit of overhead that could otherwise be avoided.

Reply:
>In our future design, it's the operator that will compute it based on the host names

Will these changes be included in this PR or a follow up?

Reply:
Follow up PR, in this PR Medusa computes the mappings, including the in place/remote nature.